### PR TITLE
feat: Add ability to configure UI through configuration

### PR DESCRIPTION
**Pull request by HypeMC**
| Q             | A                                                                                                                         |
|---------------|---------------------------------------------------------------------------------------------------------------------------|
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -

Adds the ability to configure the UI through configuration:

```yaml
nelmio_api_doc:
  html_config:
    assets_mode: bundle
    redocly_config:
      expandResponses: '200,201'
      hideDownloadButton: true
    swagger_ui_config:
      deepLinking: true
```

### CODE REVIEW

File: templates/SwaggerUi/index.html.twig [hunk 1]
```diff
@@ -75,10 +75,9 @@ file that was distributed with this source code. #}
 
     {% block swagger_initialization %}
         <script type="text/javascript">
-            (function () {
-                var swaggerUI = {{ swagger_ui_config|json_encode(65)|raw }};
-                window.onload = loadSwaggerUI(swaggerUI);
-            })();
+            window.onload = () => {
+                loadSwaggerUI({{ swagger_ui_config|json_encode(65)|raw }});
+            };
```
Comment (HypeMC) [inline templates/SwaggerUi/index.html.twig:80]: This made no sense as the `loadSwaggerUI()` function doesn't return anything.

File: templates/Redocly/index.html.twig [hunk 2]
```diff
@@ -35,4 +35,13 @@
...
\ No newline at end of file
+
+    {% block swagger_initialization %}
+        <script type="text/javascript">
+            window.onload = () => {
+                loadRedocly({{ redocly_config|json_encode(81)|raw }});
+            };
+        </script>
+    {% endblock swagger_initialization %}
+</body>
```
Comment (HypeMC) [inline templates/Redocly/index.html.twig:46]: The closing body tag was missing.

File: src/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -60,6 +60,10 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->defaultValue(['json'])
                     ->prototype('scalar')->end()
                 ->end()
+                ->arrayNode('html_config')
+                    ->info('SwaggerUI/Redocly configuration options')
+                    ->ignoreExtraKeys(false)
```
Comment (DjordyKoert) [inline src/DependencyInjection/Configuration.php:65]: Would it make sense to separate the configurations for SwaggerUI & Redocly? It is technically possible to use both UI's on different paths

Comment (HypeMC) [reply src/DependencyInjection/Configuration.php]: Good point, done

File: src/Render/Html/HtmlOpenApiRenderer.php [hunk 4]
```diff
@@ -23,13 +23,15 @@ class HtmlOpenApiRenderer implements OpenApiRenderer
...
+    private array $htmlConfig;
 
-    public function __construct($twig)
+    public function __construct($twig, array $htmlConfig)
     {
         if (!$twig instanceof \Twig_Environment && !$twig instanceof Environment) {
             throw new \InvalidArgumentException(sprintf('Providing an instance of "%s" as twig is not supported.', get_class($twig)));
         }
         $this->twig = $twig;
+        $this->htmlConfig = $htmlConfig + ['swagger_ui' => [], 'redocly' => []];
```
Comment (DjordyKoert) [inline src/Render/Html/HtmlOpenApiRenderer.php:34]:
```suggestion
        $this->htmlConfig = $htmlConfig;
```

File: src/Render/Html/HtmlOpenApiRenderer.php [hunk 5]
```diff
@@ -50,6 +53,7 @@ public function render(OpenApi $spec, array $options = []): string
                 [
                     'swagger_data' => ['spec' => json_decode($spec->toJson(), true)],
                     'assets_mode' => $options['assets_mode'],
+                    'redocly_config' => $options['redocly_config'] + $this->htmlConfig['redocly'],
```
Comment (DjordyKoert) [inline src/Render/Html/HtmlOpenApiRenderer.php:56]:
```suggestion
                    'redocly_config' => $options['redocly_config'],
```

File: src/Render/Html/HtmlOpenApiRenderer.php [hunk 6]
```diff
@@ -59,7 +63,7 @@ public function render(OpenApi $spec, array $options = []): string
             [
                 'swagger_data' => ['spec' => json_decode($spec->toJson(), true)],
                 'assets_mode' => $options['assets_mode'],
-                'swagger_ui_config' => $options['swagger_ui_config'],
+                'swagger_ui_config' => $options['swagger_ui_config'] + $this->htmlConfig['swagger_ui'],
```
Comment (DjordyKoert) [inline src/Render/Html/HtmlOpenApiRenderer.php:66]:
```suggestion
                'swagger_ui_config' => $options['swagger_ui_config'],
```

File: src/Render/Html/HtmlOpenApiRenderer.php [hunk 7]
```diff
@@ -42,6 +44,7 @@ public function render(OpenApi $spec, array $options = []): string
         $options += [
```
Comment (DjordyKoert) [inline src/Render/Html/HtmlOpenApiRenderer.php:44]:
```suggestion
        $options += $this->htmlConfig + [
```

File: src/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -55,6 +55,17 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->defaultValue(['json'])
                     ->prototype('scalar')->end()
                 ->end()
+                ->arrayNode('html_config')
+                    ->info('SwaggerUI/Redocly configuration options')
```
Comment (DjordyKoert) [inline src/DependencyInjection/Configuration.php:59]: More generic description. Furthermore this section should also include the `assets_mode` config option. https://github.com/nelmio/NelmioApiDocBundle/blob/a585e502725d7e3cfcec25f7a3d59443842006da/src/Render/Html/HtmlOpenApiRenderer.php#L43
```suggestion
                    ->info('UI configuration options')
```

**Review by DjordyKoert — CHANGES_REQUESTED** [review]
This configuration option should more generic to potentially allow more options in the future, this also aligns this new configuration option with the `DumpCommand --html-config` option and also let's it serve as a potential way to customize the default value for `--html-config`.

File: src/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -55,6 +55,17 @@ public function getConfigTreeBuilder(): TreeBuilder
...
+                    ->children()
+                        ->arrayNode('swagger_ui')
+                            ->ignoreExtraKeys(false)
+                        ->end()
+                        ->arrayNode('redocly')
+                            ->ignoreExtraKeys(false)
+                        ->end()
```
[the remark this answers is not in the document]
Comment (HypeMC) [reply src/DependencyInjection/Configuration.php]: Done

File: src/Render/Html/HtmlOpenApiRenderer.php (hunk 4 above)
Comment (HypeMC) [reply src/Render/Html/HtmlOpenApiRenderer.php]: Done

File: src/Render/Html/HtmlOpenApiRenderer.php (hunk 5 above)
Comment (HypeMC) [reply src/Render/Html/HtmlOpenApiRenderer.php]: Done

File: src/Render/Html/HtmlOpenApiRenderer.php (hunk 6 above)
Comment (HypeMC) [reply src/Render/Html/HtmlOpenApiRenderer.php]: Done

File: src/Render/Html/HtmlOpenApiRenderer.php (hunk 7 above)
Comment (HypeMC) [reply src/Render/Html/HtmlOpenApiRenderer.php]: Done

File: src/DependencyInjection/Configuration.php (hunk 8 above)
Comment (HypeMC) [reply src/DependencyInjection/Configuration.php]: Done

**Review by DjordyKoert — CHANGES_REQUESTED** [review]

File: src/Render/Html/HtmlOpenApiRenderer.php [hunk 10]
```diff
@@ -39,9 +41,10 @@ public function getFormat(): string
 
     public function render(OpenApi $spec, array $options = []): string
     {
-        $options += [
+        $options += $this->htmlConfig + [
             'assets_mode' => AssetsMode::CDN,
             'swagger_ui_config' => [],
+            'redocly_config' => [],
         ];
```
[the remark this answers is not in the document]
Comment (HypeMC) [reply src/Render/Html/HtmlOpenApiRenderer.php]: Done, I had to add `addDefaultsIfNotSet()` to make sure the keys always exist.

File: tests/DependencyInjection/NelmioApiDocExtensionTest.php [hunk 11]
```diff
@@ -296,4 +297,24 @@ public static function provideCacheConfig(): iterable
             ],
         ];
     }
+
+    public function testHtmlOpenApiRendererWithHtmlConfig()
```
[the remark this answers is not in the document]
Comment (HypeMC) [reply tests/DependencyInjection/NelmioApiDocExtensionTest.php]: Done

**Review by DjordyKoert — APPROVED** [review]

**Comment by HypeMC** [comment]
I've rebased the PR since there were conflicts.

**Comment by DjordyKoert** [comment]
> I've rebased the PR since there were conflicts.

Thank you! I was planning on doing it myself but I was not able to find the time for it yet :) I will try and have a look at everything later today if my time allows it

**Comment by DjordyKoert** [comment]
Thank you @HypeMC :)

**Pull request merged**